rust-query-crlite/src/main.rs

```diff
@@ -0,0 +1,794 @@
+    enrolledIssuers: Option<Vec<String>>,
+    incremental: bool,
+}
+
+#[derive(Deserialize)]
+struct JsonRecordAttachment {
+    hash: String,
+    location: String,
+    size: u64,
+}
+
+fn update_intermediates(int_dir: &Path) -> Result<(), CRLiteDBError> {
+    let intermediates_path = int_dir.join("crlite.intermediates");
+
+    info!("Fetching {}", ICA_LIST_URL);
+    let intermediates_bytes = &reqwest::blocking::get(ICA_LIST_URL)
+        .map_err(|_| CRLiteDBError::from("could not fetch CCADB report"))?
+        .bytes()
+        .map_err(|_| CRLiteDBError::from("could not read CCADB report"))?;
+
+    let intermediates = Intermediates::from_ccadb_csv(intermediates_bytes)
+        .map_err(|_| CRLiteDBError::from("cannot parse CCADB report"))?;
+
+    let encoded_intermediates = intermediates.encode()?;
+
+    std::fs::write(intermediates_path, &encoded_intermediates)?;
+
+    Ok(())
+}
+
+fn update_db(
+    db_dir: &Path,
+    attachment_url: &str,
+    collection_url: &str,
+) -> Result<(), CRLiteDBError> {
+    let filter_path = db_dir.join("crlite.filter");
+    let stash_path = db_dir.join("crlite.stash");
+    let enrollment_path = db_dir.join("crlite.enrollment");
+    let coverage_path = db_dir.join("crlite.coverage");
+
+    info!("Fetching remote settings data from {}", collection_url);
+    let records: RemoteSettingsData = reqwest::blocking::get(collection_url)
+        .map_err(|_| CRLiteDBError::from("could not fetch remote settings collection"))?
+        .json()
+        .map_err(|_| CRLiteDBError::from("could not read remote settings data"))?;
+
+    let (stashes, full_filters): (Vec<&JsonRecord>, Vec<&JsonRecord>) =
+        records.data.iter().partition(|x| x.incremental);
+
+    if full_filters.len() != 1 {
+        return Err(CRLiteDBError::from(
+            "multiple full filters in remote settings",
+        ));
+    }
+
+    let full_filter = full_filters[0];
+
+    let mut filter_needs_update = true;
+    let mut stash_needs_update = true;
+
+    // Skip filter update if existing filter has expected sha256 hash.
+    if filter_path.exists() {
+        let expected_digest = hex::decode(&full_filter.attachment.hash)
+            .map_err(|_| CRLiteDBError::from("full filter digest corrupted"))?;
+        let mut hasher = Sha256::new();
+        hasher.update(std::fs::read(&filter_path)?);
+        if expected_digest == hasher.finalize().as_slice() {
+            filter_needs_update = false;
+        }
+    }
+
+    // Skip stash update if the filter is fresh and the stash on disk has the expected size.
+    // (We can't easily check the provided hashes since we concatenate stashes on disk.)
+    let expected_stash_size = stashes.iter().fold(0, |x, y| x + y.attachment.size);
+    let stash_metadata = std::fs::metadata(&stash_path);
+    if !filter_needs_update
+        && stash_path.exists()
+        && stash_metadata.is_ok()
+        && stash_metadata.unwrap().len() == expected_stash_size
```

Comment:
I think you could use `map_or` or something to avoid the `is_ok` and `unwrap`.

Reply:
I left this one as is. If we `map_or` then we need a default value that is guaranteed not to equal `expected_stash_size`. We could use `u64::MAX`, but I prefer to test `is_ok`.